src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.AddProperty.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Reflection;
+using System.Text.Json.Serialization.Converters;
+
+namespace System.Text.Json.Serialization
+{
+#if MAKE_UNREVIEWED_APIS_INTERNAL
+    internal
+#else
+    public
+#endif
+    partial class JsonClassInfo
+    {
+        private void AddProperty(Type propertyType, PropertyInfo propertyInfo, Type classType, JsonSerializerOptions options)
+        {
+            JsonPropertyInfo jsonInfo = null;
+            Type collectionElementType = null;
+
+            ClassType propertyClassType = GetClassType(propertyType);
+
+            if (propertyClassType == ClassType.Value)
+            {
+                bool isNullable = (propertyType.IsGenericType && propertyType.GetGenericTypeDefinition() == typeof(Nullable<>));
+
+                if (!isNullable)
+                {
+                    // More common types are listed first.
+                    if (propertyType == typeof(string))
```

Comment:
Consider instead switching on TypeCode.

Reply:
I re-wrote that code and it now uses an array based on TypeCode